### ISSUE TITLE
Full adaption mode

### ISSUE THREAD

**Issue by YJonmo** [CVLAB-Unibo/Real-time-self-adaptive-deep-stereo#34]
Hi there,

Thanks again for this work and making it publicly available.

I have a question about the adaptation mode. 

I have stereo endoscope images without the ground truth. So I want to use them for the adaption mode. To train the network, using Blender, I created an environment to produce the stereo images and their corresponding depth maps. I used them for training your network (MADNet). 
I noticed that in the Full adaption mode, right after training, if I feed the Blender stereo images (which were used for the training) to your network, then the resulting depth map is looking good. But after a few hundreds of iteration in the Full mode, the depth maps start to look flat and the variance becomes near zero. Theoretically if I use the same images for the adaption that I used the for training, then the network should perform the same if not better. But the Full mode makes the network fail. I attached two point clouds with 90degree angle to be able to see the drop in the variance of the depth map after adaption mode on 100 images. 

Left image is test result right after training the network before running the adaption mode and right image is the same images after a 100 images (from the training) used for Full mode adaption.

![Screenshot from 2019-12-03 14-18-12](https://user-images.githubusercontent.com/12040950/70020016-17d1d180-15d7-11ea-9c67-a776ed158d11.png)

The interesting point is that if I use my own stereo endoscope images for the Full mode adaptation, then after a few hundred iterations the result start to look good. So I am not sure pretraining the network using the Blender images make any improvement in depth estimation compared with if I just had used the unsupervised approach such as Godard 2017.

**Comment by AlessioTonioni** [comment]
> Theoretically if I use the same images for the adaption that I used the for training, then the network should perform the same if not better.

Not really because the loss used for the adaptation is far from being perfect and is actually very noisy. So what you are experiencing is completely reasonable. When you train with gt you have perfect supervision and no noise in the training process, when you use the self-supervised loss you inject noise in the training process and the performance starts to decrease. Self-supervision makes sense only if you don't have any other source of ground truth. 

> The interesting point is that if I use my own stereo endoscope images for the Full mode adaptation, then after a few hundred iterations the result starts to look good.

Again this is expected because you deploy the network on completely different images, so the initial performance is quite bad. Even if the self-supervised loss is very noisy it will still provide reasonable training signals that should improve the prediction from really bad to something better.

> So I am not sure pretraining the network using the Blender images make any improvement in depth estimation compared with if I just had used the unsupervised approach such as Godard 2017.

In my experience, the pretraining helps to converge faster once you start to train in a new environment. Think of it like pretraining a classification network on imagenet and then fine-tuning for a different dataset. If you start from scratch on real images without any kind of precise supervision I think you will get worse performance, but you can definitely try it.

**Comment by YJonmo** [comment]
Thanks for the feedback!

By the way I am also trying to get the pose of the camera in the environment. 
Is there any approach to to find the camera pose similar to your approach of pretraining on ground-truth and then adapting to real stereo images without the pose?
The Monodepth2 by Godard estimates the pose but it is unsupervised. 

Thanks

**Comment by AlessioTonioni** [comment]
As far as I know no, but the simplest thing would be to take something similar to the pose estimation network from Monodepth2 by Godard and add a supervised loss on the pose of the camera. Assuming that you have groundtruth for the pose of the camera in a trainign set.

**Comment by YJonmo** [comment]
thank you
